### PR TITLE
[WIP] Upload via email

### PR DESCRIPTION
**Pull request by reefdog**
As oft requested (#396): the ability to upload documents to your account via email. Users generate a unique mailbox address for each email address they send from. Attachments are limited to 30MB by SES. Mailboxes can be revoked at any time.

- Page for viewing your mailbox addresses, adding new ones, and revoking old ones
- Ingestion pipeline using SES and Lambda to pluck off attachments and hand them off to Cloud Crowd
- Bit of a piggyback: new UI styling for this mailbox page and for some other static pages (error pages, login/reset pages) to demonstrate the new design language and direction

### CODE REVIEW

**Comment by anthonydb** [comment]
Well, this looks amazing. Kudos.

**Comment by reefdog** [comment]
@anthonydb Thaaanks!

**Comment by jashkenas** [comment]
Very exciting.

File: app/controllers/accounts_controller.rb [hunk 1]
```diff
@@ -1,60 +1,62 @@
 # The AccountsController is responsible for account management -- adding users
 # activating accounts, updating email addresses and the like.
 class AccountsController < ApplicationController
```
Comment (esthervillars) [inline app/controllers/accounts_controller.rb]: ++1 lots of cleanup!

File: app/controllers/accounts_controller.rb [hunk 2]
```diff
@@ -117,41 +120,62 @@ def update
...
+    LifecycleMailer.login_instructions(account, current_organization, current_account).deliver_now
     json nil
   end
   
   def mailboxes
-    # code goes here.
+    # TODO: Filter by status, once that exists
+    @mailboxes = UploadMailbox.where(membership_id: @current_account.memberships.pluck(:id))
+    set_minimal_nav text: 'Back to workspace', link: '/search/'
   end
```
Comment (reefdog) [inline app/controllers/accounts_controller.rb]: Good question. I went with `@current_account` because it's set via this path:

1. [`before_action :login_required`](https://github.com/documentcloud/documentcloud/blob/49f7a07032ee478353a536b2c59305f93a76c7f1/app/controllers/accounts_controller.rb#L7)
2. [ApplicationController#logged_in?](https://github.com/documentcloud/documentcloud/blob/53c5e293ddacaf88704811f511dee67ca65af4e3/app/controllers/application_controller.rb#L175-L177)
3. [ApplicationController#current_account](https://github.com/documentcloud/documentcloud/blob/53c5e293ddacaf88704811f511dee67ca65af4e3/app/controllers/application_controller.rb#L228-L232)

But elsewhere in the controller, we use `current_account` directly. I prefer @current_account, but I suppose it's less safe? @knowtheory any preference?

File: app/controllers/accounts_controller.rb [hunk 3]
```diff
@@ -117,41 +120,62 @@ def update
...
   
   def create_mailbox
+    @mailbox = UploadMailbox.create(membership: @current_account.memberships.default.first, sender: params[:email])
+    if @mailbox.valid?
```
Comment (reefdog) [inline app/controllers/accounts_controller.rb]: It's not, no. Should/need it be?

File: app/controllers/accounts_controller.rb (hunk 2 above)
Comment (jashkenas) [inline app/controllers/accounts_controller.rb]: In general, `current_account` should be preferred. It's just an accessor if it's already been looked up.

More importantly, if something makes account lookup change in the future — you only have one place to change the logic, which is not the case if you access the instance variable directly.

File: app/controllers/accounts_controller.rb (hunk 3 above)
Comment (knowtheory) [inline app/controllers/accounts_controller.rb]: we need consistency one way or the other.  I'm not 100% whether SES enforces downcasing in email headers or not.  I'd vote for yes we should downcase everything

**Comment by knowtheory** [comment]
Alllllright.  I've wrestled Lambda into submission (and will figure out where to put the code up).

For those of you with staging, you can test it out by going to https://staging.documentcloud.org/accounts/mailboxes and following the instructions there.

**Comment by knowtheory** [comment]
Oh, so there's some discussion to be had around edge cases.  If you email documents in... I think it's fair to assume that feedback can/should come via email.

With that assumption i already have it configured to set the email notification for when the upload completes.

But in addition to the desired/successful outcome... let's talk about how to handle failure mode notification:

1. Email is sent from unauthorized address to a valid mailbox (who should be notified in this circumstance?)
2. Email is sent but has no attachments

hmmm i would ask for thoughts about other edge cases, but perhaps i should sketch out the architecture of how this is wired up first :)

**Comment by reefdog** [comment]
> 1. Email is sent from unauthorized address to a valid mailbox (who should be notified in this circumstance?)

A nonexistent mailbox and a mismatch between sender/mailbox should be treated the same so as not to leak confirmation of mailbox's existence. Reply to the sender with "Sorry, we can't match your sending address to this mailbox. Go to your Mailboxes page to make sure you're sending to the correct mailbox for this sending address." Later, we can add security heuristics that notify us (and/or proper mailbox owners) if a mailbox is being hammered maliciously.

**Comment by reefdog** [comment]
Also, we should probably send these messages to the From address (i.e., whatever we're matching against `UploadMailbox#sender`), not the email's Reply-To address. Otherwise, if I'm trying to maliciously upload into famous@journo.org's DC account, then I'd send messages as `From: famous@journo.org; Reply-To: sneaky@hacker.net` so that I got any errors/notifications instead of the target account.

**Comment by reefdog** [comment]
Some email clients now move large attachments to cloud storage and send a link in the email body instead. We should be smart enough to pluck the URLs out and fetch them. Of note: some clients (but sadly not Gmail) put the URL into a `Remote-Attachment-Url` header, saving us the trouble of parsing the body.

**Comment by knowtheory** [comment]
Deployed & tested this against staging.  Confirmation that it works as intended, so i'd like to get this out into the hands of some users.
